### PR TITLE
Feature/add email verification

### PR DESCRIPTION
**Pull request by andytrue7**

### CODE REVIEW

File: client/src/services/requestService.js [hunk 1]
```diff
@@ -15,6 +15,13 @@ axios.interceptors.request.use(
     error => Promise.reject(error)
 );
 
+axios.interceptors.response.use(
+    res => res,
+    err => {
+        throw new Error(err.response.data.error.message);
```
Comment (lenchvolodymyr) [inline client/src/services/requestService.js:21]: make it safety err.?response?.data?.error?.message || err

and rename `err` to `error` and `res` to `response`

File: client/src/components/auth/VerifiedEmailComponent.vue [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    name: 'VerifiedEmailComponent',
+
+    methods: {
+        ...mapActions('auth', {
+            verifyEmail: actions.VERIFY_EMAIL
+        })
+    },
+
+    async created() {
+        await this.verifyEmail(this.$route.query);
```
Comment (lenchvolodymyr) [inline client/src/components/auth/VerifiedEmailComponent.vue:18]: add try {} catch and if error appears show notification

File: client/src/components/auth/VerifiedEmailComponent.vue [hunk 3]
```diff
@@ -0,0 +1,24 @@
+<template>
+    <div></div>
```
Comment (lenchvolodymyr) [inline client/src/components/auth/VerifiedEmailComponent.vue:2]: add localization with text: "Your account is being activated..." and if success "Your account has been activated. Please, log in" and "log in" make as link to log in page

File: client/src/components/auth/VerifiedEmailComponent.vue [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+    methods: {
+        ...mapActions('auth', {
+            verifyEmail: actions.VERIFY_EMAIL
+        })
+    },
+
+    async created() {
+        await this.verifyEmail(this.$route.query);
+        this.$router.push({ name: 'SignIn' });
```
Comment (lenchvolodymyr) [inline client/src/components/auth/VerifiedEmailComponent.vue:19]: Instead of redirect, display success message and link to log in page, as I wrote above

**Review by lenchvolodymyr — CHANGES_REQUESTED** [review]

File: api/app/Actions/Auth/EmailVerificationAction.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    public function execute(EmailVerificationRequest $request): JsonResponse
+    {
+        $user = $this->userRepository->getById($request->getId());
+
+        if (!hash_equals((string) $request->getHash(), sha1($user->getEmailForVerification()))) {
+            return response()->json([
+                'message' => 'Unauthorized'
+            ]);
```
Comment (aLLeXUs) [inline api/app/Actions/Auth/EmailVerificationAction.php:24]: Action must not generate an HTTP response. An exception should be thrown here.

File: api/app/Actions/Auth/EmailVerificationAction.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        }
+
+        if (!$user->hasVerifiedEmail()) {
+            $this->userRepository->markUserEmail($user);
+        } else {
+            return response()->json([
+                'message' => 'User already verified!'
+            ]);
```
Comment (aLLeXUs) [inline api/app/Actions/Auth/EmailVerificationAction.php:32]: The same here.

File: api/app/Actions/Auth/EmailVerificationAction.php [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        }
+
+        return response()->json([
+            'message' => 'Email verified successfully!'
+        ]);
```
Comment (aLLeXUs) [inline api/app/Actions/Auth/EmailVerificationAction.php:37]: Such response is not needed here. The errors will be handled via exceptions and if there was not any the response can be treated as successful.

File: api/app/Actions/Auth/LoginAction.php [hunk 8]
```diff
@@ -2,13 +2,28 @@
...
+    {
+        $this->userRepository = $userRepository;
+    }
+
     public function execute(LoginRequest $request): AuthenticationResponse
     {
+        $user = $this->userRepository->getByVerifiedEmail($request->getEmail());
+
+        if (!$user) {
+            throw new AuthenticationException('Email is not verified');
```
Comment (aLLeXUs) [inline api/app/Actions/Auth/LoginAction.php:24]: You can create a separate exception `NotVerifiedEmailException` that is based on AuthenticationException.

File: api/app/Http/Controllers/Api/Auth/EmailVerificationController.php [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+class EmailVerificationController extends Controller
+{
+    public function verify(Request $request, EmailVerificationAction $action)
+    {
+        $emailVerificationRequest = new EmailVerificationRequest(
+            $request->id,
+            $request->hash
+        );
+
+        return $action->execute($emailVerificationRequest);
```
Comment (aLLeXUs) [inline api/app/Http/Controllers/Api/Auth/EmailVerificationController.php:19]: You should not directly return the response from the action to the client. In this case, the response is not needed so you can return empty one.

File: api/app/Notifications/VerifyNotification.php [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+            'hash' => sha1($notifiable->getEmailForVerification()),
+        ];
+
+        $url = $this->verificationUrl($notifiable);
+
+        return (new MailMessage())
+            ->subject(Lang::get('Verify Email Address'))
+            ->line(Lang::get('Please click the button below to verify your email address.'))
+            ->action(Lang::get('Verify Email Address'), $url)
+            ->line(Lang::get('If you did not create an account, no further action is required.'));
```
Comment (aLLeXUs) [inline api/app/Notifications/VerifyNotification.php:40]: Please create a separate class for the email so the content generation will be there, and here you will only need to create the instance (example `api/app/Mail/EventCreatedMail.php`).

**Review by aLLeXUs — CHANGES_REQUESTED** [review]

File: client/src/components/auth/VerifiedEmailComponent.vue [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+        showMessage(message, type = '') {
+            this.alert.message = message;
+            this.alert.type = type;
+            this.alert.visible = true;
+        }
+    },
+
+    async created() {
+        setTimeout(() => (this.textVisible = false), 1000);
```
Comment (lenchvolodymyr) [inline client/src/components/auth/VerifiedEmailComponent.vue:54]: yod don't need timeout here. Set it false after await this.verifyEmail(this.$route.query);

**Review by lenchvolodymyr — CHANGES_REQUESTED** [review]

File: client/src/services/requestService.js (hunk 1 above)
Comment (mv-v-7) [reply client/src/services/requestService.js]: What will happen if backend is unavailable? Can we propose way to distinguish  like net::ERR_CONNECTION_REFUSED exception  it is created by Axios from  exception that we throw in backend by hand.

**Review by lenchvolodymyr — APPROVED** [review]

**Review by aLLeXUs — APPROVED** [review]

**Pull request merged**